File: php/122.php
<?php

class Solution {

    /**
     * @param Integer[] $prices
     * @return Integer
     */
    function maxProfit($prices) {
        if (empty($prices)) return 0;

        $dp = [];
        $dp[0][0] = 0;
        $dp[0][1] = -$prices[0];
        for ($i = 1; $i < count($prices); $i++) {
            $dp[$i][0] = max($dp[$i - 1][0], $dp[$i - 1][1] + $prices[$i]);
            $dp[$i][1] = max($dp[$i - 1][1], $dp[$i - 2][0] - $prices[$i]);
        }
        return $dp[count($prices) - 1][0];
    }
}

$solution = new  Solution();
var_dump($solution->maxProfit([7, 1, 5, 3, 6, 4]));
//$dp[i][k][0] = max($dp[i-1][k][0],$dp[i-1][k][1] + prices[i])
//$dp[i][k][1] = max($dp[i-1][k][1],$dp[i-1][k-1][0] - prices[i])

File: go/dp/122.go
package main

func max(x int, y int) int {
	if x > y {
		return x
	}
	return y
}

func maxProfit(prices []int) int {
	length := len(prices)
	if length == 0 {
		return 0
	}

	dp := make([][]int, length)
	dp[0] = []int{0, -prices[0]}
	for i := 1; i < length; i++ {
		dp[i] = []int{max(dp[i-1][0], dp[i-1][1]+prices[i]), max(dp[i-1][1], dp[i-1][0]-prices[i])}
	}

	return max(dp[length-1][0], dp[length-1][1])
}
